src/MongoDB.Driver.Core/Core/Operations/CountDocumentsOperation.cs

```diff
@@ -0,0 +1,238 @@
+            Ensure.IsNotNull(binding, nameof(binding));
+            using (var channelSource = binding.GetReadChannelSource(cancellationToken))
+            using (var channel = channelSource.GetChannel(cancellationToken))
+            using (var channelBinding = new ChannelReadBinding(channelSource.Server, channel, binding.ReadPreference, binding.Session.Fork()))
+            {
+                var operation = CreateOperation();
+                var cursor = operation.Execute(channelBinding, cancellationToken);
+                var result = cursor.ToList(cancellationToken);
+                return ExtractCountFromResult(result);
+            }
+        }
+
+        /// <inheritdoc/>
+        public async Task<long> ExecuteAsync(IReadBinding binding, CancellationToken cancellationToken)
+        {
+            Ensure.IsNotNull(binding, nameof(binding));
+            using (var channelSource = await binding.GetReadChannelSourceAsync(cancellationToken).ConfigureAwait(false))
+            using (var channel = await channelSource.GetChannelAsync(cancellationToken).ConfigureAwait(false))
+            using (var channelBinding = new ChannelReadBinding(channelSource.Server, channel, binding.ReadPreference, binding.Session.Fork()))
+            {
+                var operation = CreateOperation();
+                var cursor = await operation.ExecuteAsync(channelBinding, cancellationToken).ConfigureAwait(false);
+                var result = await cursor.ToListAsync(cancellationToken).ConfigureAwait(false);
+                return ExtractCountFromResult(result);
+            }
+        }
+
+        // private methods
+        private AggregateOperation<BsonDocument> CreateOperation()
+        {
+            var pipeline = CreatePipeline();
+            var operation = new AggregateOperation<BsonDocument>(_collectionNamespace, pipeline, BsonDocumentSerializer.Instance, _messageEncoderSettings)
+            {
+                Collation = _collation,
+                Hint = _hint,
+                MaxTime = _maxTime,
+                ReadConcern = _readConcern
+            };
+            return operation;
+        }
+
+        private List<BsonDocument> CreatePipeline()
+        {
+            var pipeline = new List<BsonDocument>();
+            pipeline.Add(new BsonDocument("$match", _filter ?? new BsonDocument()));
+            if (_skip.HasValue)
+            {
+                pipeline.Add(new BsonDocument("$skip", _skip.Value));
+            }
+            if (_limit.HasValue)
+            {
+                pipeline.Add(new BsonDocument("$limit", _limit.Value));
+            }
+            pipeline.Add(new BsonDocument("$group", new BsonDocument { { "_id", BsonNull.Value }, { "n", new BsonDocument("$sum", 1) } }));
+            return pipeline;
+        }
+
+        private long ExtractCountFromResult(List<BsonDocument> result)
+        {
+            switch (result.Count)
+            {
+                case 0:
+                    return 0;
```

Comment:
This is an interesting special case... when the result is zero the aggregate command returns no documents.

Seeing no documents is how you tell that the result is zero.